lib/datasets/aozora-bunko.rb

```diff
@@ -0,0 +1,164 @@
+      :person_family_name_reading,
+      :person_first_name_reading,
+      :person_family_name_collation,
+      :person_first_name_collation,
+      :person_family_name_roman,
+      :person_first_name_roman,
+      :person_type,
+      :person_birthday,
+      :person_death_day,
+      :copyright_for_person,
+      # 底本名1,底本出版社名1,底本初版発行年1,入力に使用した版1,校正に使用した版1,底本の親本名1,底本の親本出版社名1,底本の親本初版発行年1,
+      :original_book_name1,
+      :original_book_publisher_name1,
+      :original_book_first_published_date1,
+      :used_version_for_registration1,
+      :used_version_for_proofreading1,
+      :base_of_original_book_name1,
+      :base_of_original_book_publisher_name1,
+      :base_of_original_book_first_published_date1,
+      # 底本名2,底本出版社名2,底本初版発行年2,入力に使用した版2,校正に使用した版2,底本の親本名2,底本の親本出版社名2,底本の親本初版発行年2,
+      :original_book_name2,
+      :original_book_publisher_name2,
+      :original_book_first_published_date2,
+      :used_version_for_registration2,
+      :used_version_for_proofreading2,
+      :base_of_original_book_name2,
+      :base_of_original_book_publisher_name2,
+      :base_of_original_book_first_published_date2,
+      # 入力者,校正者,
+      :registered_person_name,
+      :proofreader_name,
+      # テキストファイルURL,テキストファイル最終更新日,テキストファイル符号化方式,テキストファイル文字集合,テキストファイル修正回数,
+      :text_file_url,
+      :latest_text_file_updated_date,
+      :text_file_character_encoding,
+      :text_character_set,
+      :text_file_updating_count,
+      # XHTML/HTMLファイルURL,XHTML/HTMLファイル最終更新日,XHTML/HTMLファイル符号化方式,XHTML/HTMLファイル文字集合,XHTML/HTMLファイル修正回数
+      :html_file_url,
+      :latest_html_file_updated_date,
+      :html_file_character_encoding,
+      :html_character_set,
+      :html_file_updating_count
+    )
+
+    class Record
+      def text
+        return @text unless @text.nil?
+        return '' if text_file_url.nil? || text_file_url.empty?
+
+        # when url is not zip file, it needs to open web page by brower and has to download
+        # e.g. https://mega.nz/file/6tMxgAjZ#PglDDyJL0syRhnULqK0qhTMC7cktsgqwObj5fY_knpE
+        return '' unless text_file_url.end_with?('.zip')
+
+        downloader = Datasets::Downloader.new(text_file_url)
+        downloader.download(output_path)
+
+        Zip::File.open(output_path) do |zip_file|
+          zip_file.each do |entry|
+            next unless entry.file?
+
+            entry.get_input_stream do |stream|
+              @text = stream.read.encode(Encoding::UTF_8, Encoding::SHIFT_JIS)
+            end
+          end
+        end
+
+        FileUtils.rmtree(output_path.to_s, secure: true) if output_path.exist?
+
+        @text
+      end
+
+      private
+
+      def output_path
+        base_dir = case RUBY_PLATFORM
+                   when /mswin/, /mingw/
+                     ENV['LOCALAPPDATA'] || '~/AppData/Local'
+                   when /darwin/
+                     '~/Library/Caches'
+                   else
+                     ENV['XDG_CACHE_HOME'] || '~/.cache'
+                   end
```

Comment:
Umm. We need to introduce some reusable code for cache directory and Zip processing like existing `Downloader`...
Do you want to work on it? Or can you wait for implemented by me?

Reply:
I wanted to reuse code of  cache and downloading , but it was difficult for me  because they are implemented inside `Dataset` class.

> Do you want to work on it?

Yes, i want to try it.

Reply:
Great! Feel free to ask me if you have any concerns!

Reply:
- `Datasets#cash_dir_path` は多くのクラスで使われているのでそのAPIは変えたくない
- `cache_dir_path` , `clear_cache!` を分離

ということをまずは実現するために以下のようにしました。

```ruby
module Datasets
  class Cache
    def initialize(dir)
      @dir = dir
    end

    def dir_path
      base_dir = case RUBY_PLATFORM
                 when /mswin/, /mingw/
                   ENV['LOCALAPPDATA'] || '~/AppData/Local'
                 when /darwin/
                   '~/Library/Caches'
                 else
                   ENV['XDG_CACHE_HOME'] || '~/.cache'
                 end
      Pathname(base_dir).expand_path + 'red-datasets' + @dir
    end

    def clear!
      FileUtils.rmtree(dir_path.to_s, secure: true) if dir_path.exist?
    end
  end
```

```ruby
module Datasets
  class Dataset

    def clear_cache!
      @cache ||= Cache.new(@metadata.id)
      @cache.clear!
    end

    private

    def cache_dir_path
      @cache ||= Cache.new(@metadata.id)
      @cache.dir_path
    end
  end
end
```


ただ、 `Cache` としてしまうと `Cash#set('key', 'value')` `Cash#get('key')` で使うイメージがあるので、どうしたもののか、と思っています。

Reply:
たしかに既存の`Datasets::Dataset`のAPIは変えたくないですね！さすがです！

これでどうですか！？

```ruby
module Datasets
  class CachePath
    def initialize(id)
      @id = id
    end

    def base_dir
      Pathname(system_cache_dir).expand_path + 'red-datasets' + @id
    end

    def clear!
      FileUtils.rmtree(base_dir.to_s, secure: true) if base_dir.exist?
    end

    private
    def system_cache_dir
      case RUBY_PLATFORM
      when /mswin/, /mingw/
        ENV['LOCALAPPDATA'] || '~/AppData/Local'
      when /darwin/
        '~/Library/Caches'
      else
        ENV['XDG_CACHE_HOME'] || '~/.cache'
      end
    end
  end
end
```

```ruby
module Datasets
  class Dataset
    def clear_cache!
      cache_path.clear!
    end

    private
    def cache_path
      @cache_path ||= CachePath.new(@metadata.id)
    end

    def cache_dir_path
      cache_path.base_dir
    end
  end
end
```

Reply:
ありがとうございます。いいと思います。
名前と機能が一致していてスッキリしました。
名前大事ですね。